src/babel-serializer/geometry/off.cc

```diff
@@ -0,0 +1,345 @@
+        {
+            if (*curr == '#')
+                next_line();
+            skip_spaces();
+        }
+    };
+    auto const next_token = [&]() -> cc::string_view {
+        skip_spaces();
+        if (*curr == '#' || cc::is_space(*curr))
+        {
+            // error, unexpected end of line
+            return {};
+        }
+        auto const start = curr;
+        while (curr != end && !cc::is_space(*curr) && *curr != '#')
+            ++curr;
+        return cc::string_view(start, curr);
+    };
+
+    if (curr == end)
+    {
+        on_error(data, data, "Faile to parse OFF: File empty", severity::error);
+        return {};
+    }
+
+    // parse header, if there is one
+    {
+        skip_spaces_and_comments();
+        if (curr < end - 1 && curr[0] == 'S' && curr[1] == 'T')
+        {
+            has_tex_coords = true;
+            curr += 2;
+        }
+        if (curr != end && curr[0] == 'C')
+        {
+            has_colors = true;
+            ++curr;
+        }
+        if (curr != end && curr[0] == 'N')
+        {
+            has_normals = true;
+            ++curr;
+        }
+        if (curr != end && curr[0] == '4')
+        {
+            has_homogeneous = true;
+            ++curr;
+        }
+        if (curr != end && curr[0] == 'n')
+        {
+            has_dimension = true;
+            ++curr;
+        }
+        if (curr < end - 2 && curr[0] == 'O' && curr[1] == 'F' && curr[2] == 'F')
+        {
+            has_header = true;
+            curr += 3;
+        }
+        if (has_header)
+        {
+            if (has_dimension) // must be explicityly given if there is no header
+            {
+                next_line();
+                if (!cc::from_string(next_token(), dimension))
+                {
+                    on_error(data, data.subspan(curr - data_as_string_view.begin()),
+                             "Failed to parse OFF file: Header contained dimension hint 'n' but no dimension information was present", severity::error);
+                    dimension = 3;
+                }
+            }
+            else
+            {
+                dimension = 3;
+            }
+            if (has_homogeneous)
+            {
+                ++dimension;
+            }
+            if (dimension > 4)
+            {
+                on_error(data, data.subspan(curr - data_as_string_view.begin()), "Failed to parse OFF file: At most 4 dimensional points supported",
+                         severity::error);
+            }
+        }
+        else
+        {
+            // no header given, reset
+            curr = data_as_string_view.begin();
+            has_tex_coords = cfg.has_tex_coords;
+            has_normals = cfg.has_normals;
+            has_colors = cfg.has_colors;
+            has_dimension = false;
+            dimension = cfg.dimension;
+        }
+    }
+
+    int n_vertices = 0;
+    int n_faces = 0;
+    int n_edges = 0;
+    {
+        next_line();
```

Comment:
a warning if the first line does not contain `OFF`?

Reply:
From what I can tell, it seems that is quite common.

Reply:
damn.